For further reference, CAN-2005-0605 has been assigned to this issue.
See https://bugs.freedesktop.org/show_bug.cgi?id=1920 for more info.
diff -Naur xc.orig/extras/Xpm/lib/create.c xc/extras/Xpm/lib/create.c
--- xc.orig/extras/Xpm/lib/create.c	2004-12-17 02:09:36.000000000 +0100
+++ xc/extras/Xpm/lib/create.c	2005-03-17 09:52:24.000000000 +0100
@@ -1215,10 +1215,10 @@
     register char *src;
     register char *dst;
     register unsigned int *iptr;
-    register unsigned int x, y, i;
+    register unsigned int x, y;
     register char *data;
     Pixel pixel, px;
-    int nbytes, depth, ibu, ibpp;
+    int nbytes, depth, ibu, ibpp, i;
 
     data = image->data;
     iptr = pixelindex;
diff -Naur xc.orig/extras/Xpm/lib/scan.c xc/extras/Xpm/lib/scan.c
--- xc.orig/extras/Xpm/lib/scan.c	2004-12-17 02:09:36.000000000 +0100
+++ xc/extras/Xpm/lib/scan.c	2005-03-17 09:52:24.000000000 +0100
@@ -621,8 +621,8 @@
     char *dst;
     unsigned int *iptr;
     char *data;
-    unsigned int x, y, i;
-    int bits, depth, ibu, ibpp, offset;
+    unsigned int x, y;
+    int bits, depth, ibu, ibpp, offset, i;
     unsigned long lbt;
     Pixel pixel, px;
 
@@ -633,6 +633,9 @@
     ibpp = image->bits_per_pixel;
     offset = image->xoffset;
 
+    if (image->bitmap_unit < 0)
+	    return (XpmNoMemory);
+
     if ((image->bits_per_pixel | image->depth) == 1) {
 	ibu = image->bitmap_unit;
 	for (y = 0; y < height; y++)
